plasma_framework/contracts/src/utils/RLPReader.sol

```diff
@@ -135,102 +178,85 @@ library RLPReader {
-     * @return The encoded RLPItem length
+     * @return The length of the RLPItem (including the length field) and the offset of the payload
      */
-    function decodeItemLengthUnsafe(uint256 memPtr) internal pure returns (uint256) {
-        uint256 decodedItemLengthUnsafe;
+    function decodeLengthAndOffset(uint256 memPtr) internal pure returns (uint256, uint256) {
+        uint256 decodedLength;
+        uint256 offset;
         uint256 byte0;
+
         // solhint-disable-next-line no-inline-assembly
         assembly {
             byte0 := byte(0, mload(memPtr))
         }
 
         if (byte0 < STRING_SHORT_START) {
-            decodedItemLengthUnsafe = 1;
+            // Item is a single byte
+            decodedLength = 1;
+            offset = 0;
         } else if (byte0 < STRING_LONG_START) {
-            decodedItemLengthUnsafe = (byte0 - STRING_SHORT_START) + 1;
+            // Item is a short string (length <= 55 bytes)
+            decodedLength = (byte0 - STRING_SHORT_START) + 1;
+            if (decodedLength == 2){
+                uint256 byte1;
+                // solhint-disable-next-line no-inline-assembly
+                assembly {
+                    byte1 := byte(0, mload(add(memPtr, 1)))
+                }
+                // A single byte below 0x80 must be encoded as itself.
+                require(byte1 >= STRING_SHORT_START, "Invalid RLP encoding");
+            }
+            offset = 1;
         } else if (byte0 < LIST_SHORT_START) {
+            // Item is a long string (length > 55 bytes)
             uint256 dataLen;
             uint256 byte1;
             // solhint-disable-next-line no-inline-assembly
             assembly {
-                let byteLen := sub(byte0, 0xb7) // # of bytes the actual length is
+                let lengthLen := sub(byte0, 0xb7) // Number of bytes the actual length is
                 memPtr := add(memPtr, 1) // skip over the first byte
+                byte1 := byte(0, mload(memPtr)) // read the first byte to check for leading zeros
 
-                byte1 := byte(0, mload(memPtr))
-
-                /* 32 byte word size */
-                dataLen := div(mload(memPtr), exp(256, sub(32, byteLen))) // right shifting to get the len
-                decodedItemLengthUnsafe := add(dataLen, add(byteLen, 1))
+                // right shift to the correct position
+                dataLen := div(mload(memPtr), exp(256, sub(WORD_SIZE, lengthLen)))
+                decodedLength := add(dataLen, add(lengthLen, 1))
             }
-            // Check valid long string i.e. value of length > MAX_SHORT_LEN with no leading zeros
+            // Check that the length has no leading zeros
             require(byte1 != 0, "Invalid RLP encoding");
+            // Check that the value of length > MAX_SHORT_LEN
             require(dataLen > MAX_SHORT_LEN, "Invalid RLP encoding");
+            // Calculate the offset
+            offset = (byte0 - (STRING_LONG_START - 1)) + 1;
```

Comment:
minor: it would easier to read if this is sth like how it is calculated previously in `add(legthLen, 1)`.

but probably just me as I am checking this function agains the RLP spec python example: https://github.com/ethereum/wiki/wiki/RLP